### PR TITLE
Cloud metadata

### LINKED ISSUE #1264: Collect cloud metadata
Implementing https://github.com/elastic/apm/issues/256 and https://github.com/elastic/apm/issues/289.
Reference implementation in [Python agent](https://github.com/elastic/apm-agent-python/pull/826).

### PR DESCRIPTION
**Pull request by eyalkoren**
## What does this PR do?
Closes #1264
Extending the work started at #1453 

## Checklist
  - [x] [Spec](https://github.com/elastic/apm/blob/master/specs/agents/metadata.md#cloud-provider-metadata) is implemented
  - [x] Updating [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] Unit tests for raw metadata deserialization logic
  - [x] Unit tests for metadata serialization logic
  - [x] Unit tests for the new async metadata creation mechanism with all `CLOUD_PROVIDER` options
  - [x] Added an API method or config option? Document in which version this will be introduced
  - [x] I have made corresponding changes to the documentation
  - [x] manual tests on each cloud with all `CLOUD_PROVIDER` options
  - [x] manual tests on each cloud with invalid URL or improper input parsing - for future changes in cloud APIs
  - [x] manual test with older APM server

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java [hunk 1]
```diff
@@ -160,10 +157,15 @@ String fetchConfig(final ConfigurationRegistry configurationRegistry) {
             return null;
         }
         try {
+            payloadSerializer.blockUntilReady();
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java:160]: Not saying this doesn't work but did you consider providing a `Future<MetaData>` in the constructor and blocking on `Future#get` here and call `DslJsonSerializer#serializeMetadata(MetaData metaData)`? That way, we could keep state management out of the `DslJsonSerializer`.

Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java:160]: Of course, this is how I started, but there were several reasons I changed that: the first is that nobody really needs the metadata object, it is spread it around just so it can be handed to serialization. 
In addition, I wanted to create a central serialization for it, meaning that `MetaData` (or something else) needs to serialize and hold the serialized data. This would mean that `MetaData` (or something else) now either create and use a `PayloadSerializer`, or I extract the metadata serialization from the rest, but that proved to be too painful and risky for this PR. Both options looked bad when I tried them.
Lastly, I don't think it makes sense to block during serialization itself, for example- you better only start the HTTP request when ready to write to it. So I thought it would be better if the serialization interface provides a good indication for that, instead of relying on `Future#get` (this is not the only usage of it).

Bottom line - there are many options to do that and I played with several ending with this. Given the multiple dependencies we had on `MetaData` and the time I wanted to spend on this, this seems as the least ugly.

Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java:160]: Thanks for explaining the background

File: apm-agent-core/src/test/java/co/elastic/apm/agent/impl/MetaDataTest.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            metaDataFuture.get(0, TimeUnit.MILLISECONDS);
+        } catch (TimeoutException e) {
+            timeoutException = e;
+        }
+        if (provider != GCP && provider != currentCloudProvider) {
+            // The GCP fails quickly on DNS lookup
+            assertThat(timeoutException).isInstanceOf(TimeoutException.class);
+        }
+        // In AWS we may need two timeouts - one for the API token and one for the metadata itself
+        long timeout = (long) (coreConfiguration.geCloudMetadataDiscoveryTimeoutMs() * ((provider == AWS) ? 2.5 : 1.5));
```
Comment (SylvainJuge) [inline apm-agent-core/src/test/java/co/elastic/apm/agent/impl/MetaDataTest.java:98]: [question] would there be a way to avoid static timeouts as those might imply test flakyness ?

File: apm-agent-core/src/test/resources/logback-test.xml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (SylvainJuge) [inline apm-agent-core/src/test/resources/logback-test.xml:1]: [question] why do we have to use `logback-test.xml` here an not `logback.xml` ?

File: apm-agent-plugins/apm-kafka-plugin/apm-kafka-headers-plugin/src/test/java/co/elastic/apm/agent/kafka/KafkaClientVersionsIT.java [hunk 4]
```diff
@@ -30,6 +30,7 @@
 import org.junit.runners.Parameterized;
 
 import java.util.Arrays;
+import java.util.List;
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-kafka-plugin/apm-kafka-headers-plugin/src/test/java/co/elastic/apm/agent/kafka/KafkaClientVersionsIT.java:33]: [minor] likely unused import

File: apm-agent-plugins/apm-log-shipper-plugin/src/main/java/co/elastic/apm/agent/log/shipper/ApmServerLogShipper.java [hunk 5]
```diff
@@ -51,12 +50,12 @@
...
 
-    public ApmServerLogShipper(ApmServerClient apmServerClient, ReporterConfiguration reporterConfiguration, MetaData metaData, PayloadSerializer payloadSerializer) {
-        super(reporterConfiguration, metaData, payloadSerializer, apmServerClient);
+    public ApmServerLogShipper(ApmServerClient apmServerClient, ReporterConfiguration reporterConfiguration, PayloadSerializer payloadSerializer) {
+        super(reporterConfiguration, payloadSerializer, apmServerClient);
     }
 
     @Override
-    public boolean onLineAvailable(TailableFile tailableFile, byte[] line, int offset, int length, boolean eol) throws IOException {
+    public boolean onLineAvailable(TailableFile tailableFile, byte[] line, int offset, int length, boolean eol) throws Exception {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-log-shipper-plugin/src/main/java/co/elastic/apm/agent/log/shipper/ApmServerLogShipper.java:58]: [question] why change to the broad `Exception` here ? If it's a way to make uses of this API to having to handle potential exceptions, that would be worth to add the the method Javadoc to be explicit.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+            }
+        }
+        String awsMetadataUrl = "http://169.254.169.254/latest/dynamic/instance-identity/document";
+        Map<String, String> documentHeaders = null;
+        if (token != null) {
+            documentHeaders = new HashMap<>(1);
+            documentHeaders.put("X-aws-ec2-metadata-token", token);
+        }
+        String metadata = executeRequest(awsMetadataUrl, "GET", documentHeaders, queryTimeoutMs);
+        logger.debug("AWS metadata retrieved");
```
Comment (SylvainJuge) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java:202]: maybe output JSON for easy debugging

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+                                    // Expected - trial and error method
+                                }
+                                return gcpInfo;
+                            }
+                        });
+                    } finally {
+                        executor.shutdown();
+                    }
+
+                    long futureTimeout = queryTimeoutMs + 200;
```
Comment (SylvainJuge) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java:138]: [question] why do we add a fixed extra timeout for the future, how do we know that 200ms is fine ?

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+                }
+                case AZURE: {
+                    try {
+                        cloudProviderInfo = getAzureMetadata(queryTimeoutMs);
+                    } catch (Exception e) {
+                        unexpectedError = e;
+                    }
+                    break;
+                }
+                case AUTO: {
```
Comment (SylvainJuge) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java:104]: [minor] maybe move this large block into a separate method as it's already nested in a switch statement.

**Review by SylvainJuge — COMMENTED** [review]
Nice! we can definitely see that having an asynchronous fetching of metadata is not without complications.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java (hunk 1 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java:160]: I don't mind, but I think it is weird to explain why something was chosen over something else you wouldn't necessarily consider as a default option. Imagine you only see this code now for the first time- `MetaData` can be easily an internal serialization thing not known to the outer world (it is not package private only not to broaden the refactoring of this PR even more). In such case, this is the interface of the serializer- it needs time to fully initialize, no matter why, so users of it should be aware. I made efforts to emphasize this in the API as much as possible, including javadoc and an Exception which is part of the API.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/DslJsonSerializer.java [hunk 9]
```diff
@@ -132,7 +141,7 @@ public void flush() throws IOException {
     }
 
     @Override
-    public void flush() throws IOException {
+    public void fullFlush() throws IOException {
```
[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/DslJsonSerializer.java:144]: It is explained in the interface, but I will copy to here as well - if you missed it others will to 🙂

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/PayloadSerializer.java [hunk 10]
```diff
@@ -71,4 +102,10 @@
     JsonWriter getJsonWriter();
 
     void writeBytes(byte[] bytes, int len);
+
+    class UninitializedException extends Exception {
```
[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/PayloadSerializer.java:106]: The main purpose of this type is to make anyone calling methods that throws it aware of the requirement for initialization, and the javadoc explains that. I prefer you notice this in compilation and for that it needs to be a checked Exception.

> it's not a recoverable

Not sure why - it may affect only the first send/serialization and not the subsequent ones.

File: apm-agent-core/src/test/java/co/elastic/apm/agent/impl/MetaDataTest.java (hunk 2 above)
Comment (eyalkoren) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/impl/MetaDataTest.java:98]: I don't think this implies flakiness, it is a **timeout** with a big safety cushion - discovery of AWS is very quick on AWS and almost exactly 2 seconds elsewhere, with Azure it would be quick on Azure and about 1 second elsewhere. It was my test confirming that all metadata discovery queries occur in parallel. If we see it is flaky, it's extremely easy to just change the timeout to 5 seconds, but a test with a timeout is necessary as we want to make sure nobody breaks the metadata discovery timeouts and let them just execute for too long.

File: apm-agent-core/src/test/resources/logback-test.xml (hunk 3 above)
Comment (eyalkoren) [reply apm-agent-core/src/test/resources/logback-test.xml:1]: We don't have to, but it is very specific for tests (the schema validation library) and it works...

File: apm-agent-plugins/apm-log-shipper-plugin/src/main/java/co/elastic/apm/agent/log/shipper/ApmServerLogShipper.java (hunk 5 above)
Comment (eyalkoren) [reply apm-agent-plugins/apm-log-shipper-plugin/src/main/java/co/elastic/apm/agent/log/shipper/ApmServerLogShipper.java:58]: Because it felt inappropriate to add `UninitializedException` (serializer-specific) to the `FileChangeListener` interface...
Would you prefer that? Or something else?

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java (hunk 6 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java:202]: I did that and got very big output, including SSH keys 😬 ...

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java (hunk 7 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java:138]: > why do we add a fixed extra timeout for the future

We want to limit this blocking somehow, but if I use the same timeout used for the HTTP query, that would be too tight, it needs to be longer.

> how do we know that 200ms is fine ?

We don't know, we live on the edge 😛

**Review by SylvainJuge — APPROVED** [review]
